GoogleDataTransport/GDTCORLibrary/GDTCORDatabase.m

```diff
@@ -0,0 +1,316 @@
+                                                BOOL *_Nonnull stop) {
+    GDTCORSQLBindObjectToParam(stmt, key.intValue, obj);
+  }];
+  BOOL result = GDTCORSQLRunQuery(db, stmt, eachRow);
+  if (cacheStmt) {
+    GDTCORSQLReset(stmt);
+  } else {
+    GDTCORSQLFinalize(stmt);
+  }
+  return result;
+}
+
+/** Runs a series of migrations, setting the user_version PRAGMA to the highest version number.
+ *
+ * @param db The db to operate on.
+ * @param migrations The map of user_version numbers to the SQL statement needed to update to that
+ *   version.
+ * @param stmtCache The statement cache.
+ */
+static void RunMigrations(sqlite3 *db,
+                          NSDictionary<NSNumber *, NSString *> *migrations,
+                          CFMutableDictionaryRef stmtCache) {
+  if (!db) {
+    GDTCORLogError(GDTCORMCEDatabaseError, @"%@", @"The database is closed");
+    return;
+  }
+  int userVersion = GetUserVersion(db);
+  __block int newUserVersion = userVersion;
+  [migrations enumerateKeysAndObjectsUsingBlock:^(NSNumber *_Nonnull key, NSString *_Nonnull obj,
+                                                  BOOL *_Nonnull stop) {
+    if (key.intValue > newUserVersion) {
+      if (RunNonQuery(db, nil, obj, stmtCache, NO)) {
+        newUserVersion = key.intValue;
+      } else {
+        GDTCORLogError(GDTCORMCEDatabaseError, @"%@", @"Migration failed: version:%@ statement:%@",
+                       key, obj);
+      }
+    }
+  }];
+
+  if (newUserVersion != userVersion) {
+    SetUserVersion(db, newUserVersion);
+  }
+}
+
+#pragma mark - GDTCORDatabase
+
+@implementation GDTCORDatabase {
+  /** The sqlite database. */
+  sqlite3 *_db;
+}
+
+- (nullable instancetype)initWithURL:(nullable NSURL *)dbFileURL
+                 migrationStatements:
+                     (nonnull NSDictionary<NSNumber *, NSString *> *)migrationStatements {
+  // Only 1 instance per file is allowed. Return nil if one is already open.
+  static dispatch_once_t onceToken;
+  dispatch_once(&onceToken, ^{
+    instancesToDBFiles = [[NSMutableDictionary alloc] init];
+  });
+  __block GDTCORDatabase *extantDB;
+  dispatch_barrier_sync(dispatch_get_global_queue(QOS_CLASS_DEFAULT, 0), ^{
+    extantDB = instancesToDBFiles[dbFileURL.path];
+  });
+  if (extantDB) {
+    return nil;
+  }
+
+  // Create an instance.
+  self = [super init];
+  if (self) {
+    NSString *dbPath = dbFileURL ? dbFileURL.path : @":memory:";
+    if (!GDTCORSQLOpenDB(&_db, dbPath)) {
+      return nil;
+    }
+    _dbQueue = dispatch_queue_create("com.google.GDTCORDatabase", DISPATCH_QUEUE_SERIAL);
+    _fileURL = dbFileURL;
+    _stmtCache =
+        CFDictionaryCreateMutable(CFAllocatorGetDefault(), 0, &kCFTypeDictionaryKeyCallBacks, NULL);
+    dispatch_async(_dbQueue, ^{
+      RunMigrations(self->_db, [migrationStatements copy], self -> _stmtCache);
+    });
+    dispatch_barrier_sync(dispatch_get_global_queue(QOS_CLASS_DEFAULT, 0), ^{
```

Comment:
I could not find an explicit statement that `dispatch_get_global_queue(QOS_CLASS_DEFAULT, 0)` returns the same queue all the times for any iOS. Would it be safer to create a local concurrent queue together with creating `instancesToDBFiles` ?

Reply:
From [the docs](https://developer.apple.com/library/archive/documentation/General/Conceptual/ConcurrencyProgrammingGuide/OperationQueues/OperationQueues.html#//apple_ref/doc/uid/TP40008091-CH102-SW5)

> The system provides each application with four concurrent dispatch queues. These queues are global to the application and are differentiated only by their priority level. Because they are global, you do not create them explicitly. Instead, you ask for one of the queues using the dispatch_get_global_queue function.

That phrasing made me think that this was fine, but then I just found this in some [different docs](https://developer.apple.com/documentation/dispatch/1452797-dispatch_barrier_async)

> The queue you specify should be a concurrent queue that you create yourself using the dispatch_queue_create function. If the queue you pass to this function is a serial queue or one of the global concurrent queues, this function behaves like the dispatch_async function.

That's not actually what I want, so I reworked this.